File: csharp/Link.Foundation.Links.Notation/ParseException.cs
using System;
using System.Text;

namespace Link.Foundation.Links.Notation
{
    /// <summary>
    /// The error raised when a document does not parse, carrying the position the parser
    /// stopped at.
    /// </summary>
    /// <remarks>
    /// The position is the furthest one the parser reached, which is the character the
    /// document stops making sense at rather than the point the last alternative gave up
    /// on. It derives from <see cref="FormatException"/>, the exception the generated
    /// parser raised on its own, so code that catches that keeps working.
    /// </remarks>
    public class ParseException : FormatException
    {
        /// <summary>The number of characters a quoted line is cut down to.</summary>
        private const int QuotedLineWidth = 80;

        /// <summary>What a message writes in place of the part of a long line it left out.</summary>
        private const string Ellipsis = "...";

        /// <summary>
        /// Creates an exception describing where <paramref name="subject"/> stopped parsing.
        /// </summary>
        /// <param name="subject">The document that failed to parse.</param>
        /// <param name="offset">Offset of the position the parser stopped at.</param>
        /// <param name="innerException">The error the generated parser raised, if any.</param>
        public ParseException(string subject, int offset, Exception? innerException = null)
            : this(Locate(subject ?? string.Empty, offset), innerException)
        {
        }

        private ParseException(Position position, Exception? innerException)
            : base(Describe(position), innerException)
        {
            Offset = position.Offset;
            Line = position.Line;
            Column = position.Column;
            Found = position.Found;
            LineText = position.LineText;
        }

        /// <summary>Offset of the offending position from the start of the document.</summary>
        public int Offset { get; }

        /// <summary>Line the offending position is on, counted from 1.</summary>
        public int Line { get; }

        /// <summary>Column the offending position is at, counted from 1.</summary>
        public int Column { get; }

        /// <summary>The character found instead, or null at the end of the document.</summary>
        public char? Found { get; }

        /// <summary>The offending line, as written, without its line ending.</summary>
        public string LineText { get; }

        /// <summary>The one-line summary: where the parser stopped and what stands there.</summary>
        public string Summary => Summarize(new Position(Offset, Line, Column, Found, LineText));

        /// <summary>
        /// The offending line with a caret under the offending column, quoted the way a
        /// compiler quotes source. A long line is shown as a window around the caret, so the
        /// message stays the same size whether the document has ten lines or fifteen hundred.
        /// </summary>
        public string Snippet => Quote(new Position(Offset, Line, Column, Found, LineText));

        /// <summary>Where the parser stopped, and what the document holds there.</summary>
        private readonly record struct Position(
            int Offset,
            int Line,
            int Column,
            char? Found,
            string LineText);

        /// <summary>The message the exception carries: the summary and the quoted line.</summary>
        private static string Describe(Position position) =>
            $"Syntax error at {Summarize(position)}\n{Quote(position)}";

        private static string Summarize(Position position)
        {
            var found = position.Found.HasValue
                ? $"\"{Escape(position.Found.Value)}\""
                : "end of input";
            return $"line {position.Line}, column {position.Column}: unexpected {found}";
        }

        private static string Quote(Position position)
        {
            var (quoted, column) = QuoteLine(position.LineText, position.Column);
            var number = position.Line.ToString();
            var gutter = new string(' ', number.Length);
            return $"{number} | {quoted}\n{gutter} | {new string(' ', column - 1)}^";
        }

        private static string Escape(char character) => character switch
        {
            '\n' => "\\n",
            '\r' => "\\r",
            '\t' => "\\t",
            '"' => "\\\"",
            '\\' => "\\\\",
            _ => character.ToString(),
        };

        /// <summary>
        /// Turns the position the parser stopped at into a line, a column and the line
        /// itself, so the message can point at the defect instead of quoting the rest of the
        /// document.
        /// </summary>
        private static Position Locate(string document, int offset)
        {
            offset = Math.Max(0, Math.Min(offset, document.Length));

            var line = 1;
            var lineStart = 0;
            for (var index = 0; index < offset; index++)
            {
                if (document[index] == '\n')
                {
                    line++;
                    lineStart = index + 1;
                }
            }

            var lineEnd = document.IndexOf('\n', lineStart);
            if (lineEnd < 0) lineEnd = document.Length;

            var lineText = document.Substring(lineStart, lineEnd - lineStart).TrimEnd('\r');
            var found = offset < document.Length ? document[offset] : (char?)null;
            return new Position(offset, line, offset - lineStart + 1, found, lineText);
        }

        /// <summary>
        /// Cuts <paramref name="line"/> down to a window around <paramref name="column"/>, and
        /// says which column the offending character sits at in that window. Both columns
        /// count from 1.
        /// </summary>
        private static (string Quoted, int Column) QuoteLine(string line, int column)
        {
            if (line.Length <= QuotedLineWidth) return (line, column);

            var target = column - 1;
            var lastStart = line.Length - QuotedLineWidth;
            var start = Math.Min(Math.Max(target - (QuotedLineWidth / 2), 0), lastStart);
            var end = start + QuotedLineWidth;

            var quoted = new StringBuilder();
            if (start > 0) quoted.Append(Ellipsis);
            quoted.Append(line, start, QuotedLineWidth);
            if (end < line.Length) quoted.Append(Ellipsis);

            var shift = start > 0 ? Ellipsis.Length : 0;
            return (quoted.ToString(), target - start + shift + 1);
        }
    }
}

File: php/src/ParseException.php
<?php

declare(strict_types=1);

namespace LinkFoundation\LinksNotation;

use Exception;
use Throwable;

/**
 * Exception raised when parsing fails.
 *
 * When the parser knows where the document stopped making sense, the
 * exception carries that position as an offset, a line and a column, and
 * quotes the offending line with a caret under the offending column, the way
 * the other Links Notation implementations do. A document nested deeper than
 * the parser's maxDepth is refused with this exception; then getMaxDepth()
 * says how deep the nesting may go.
 */
class ParseException extends Exception
{
    /** The number of characters a quoted line is cut down to. */
    private const QUOTED_LINE_WIDTH = 80;

    /** What a message writes in place of the part of a long line it left out. */
    private const ELLIPSIS = '...';

    private ?int $documentOffset = null;
    private ?int $documentLine = null;
    private ?int $documentColumn = null;
    private ?string $lineText = null;
    private ?int $maxDepth = null;

    public function __construct(string $message = '', int $code = 0, ?Throwable $previous = null)
    {
        parent::__construct($message, $code, $previous);
    }

    /**
     * The exception for a document nested deeper than $maxDepth, pointing at
     * the group or the line that is one level too deep.
     *
     * @param string $document The document being parsed
     * @param int    $offset   Byte offset of the offending position
     * @param int    $maxDepth The deepest nesting the parser accepts
     */
    public static function nestingTooDeep(string $document, int $offset, int $maxDepth): self
    {
        [$offset, $line, $column, $lineText] = self::locate($document, $offset);
        $summary = "line {$line}, column {$column}: nesting depth exceeds the maximum of {$maxDepth}";

        $error = new self("Nesting too deep at {$summary}\n" . self::quote($line, $lineText, $column));
        $error->documentOffset = $offset;
        $error->documentLine = $line;
        $error->documentColumn = $column;
        $error->lineText = $lineText;
        $error->maxDepth = $maxDepth;

        return $error;
    }

    /** Byte offset of the offending position from the start of the document, when known. */
    public function getOffset(): ?int
    {
        return $this->documentOffset;
    }

    /**
     * Line of the document the offending position is on, counted from 1, when
     * known. (Exception::getLine() is the line of PHP source that threw.)
     */
    public function getLineNumber(): ?int
    {
        return $this->documentLine;
    }

    /** Column the offending position is at, in characters, counted from 1, when known. */
    public function getColumn(): ?int
    {
        return $this->documentColumn;
    }

    /** The offending line, as written, without its line ending, when known. */
    public function getLineText(): ?string
    {
        return $this->lineText;
    }

    /** The offending line with a caret under the offending column, when known. */
    public function getSnippet(): ?string
    {
        if ($this->documentLine === null || $this->documentColumn === null || $this->lineText === null) {
            return null;
        }

        return self::quote($this->documentLine, $this->lineText, $this->documentColumn);
    }

    /**
     * The deepest nesting allowed, when the document is nested deeper; null
     * for any other error.
     */
    public function getMaxDepth(): ?int
    {
        return $this->maxDepth;
    }

    /**
     * Turn an offset into a line, a column and the line itself. CR, LF and
     * CRLF all end a line; the column counts characters, not bytes.
     *
     * @return array{0: int, 1: int, 2: int, 3: string} Offset, line, column and line text
     */
    private static function locate(string $document, int $offset): array
    {
        $offset = max(0, min($offset, strlen($document)));
        $line = 1;
        $lineStart = 0;
        $cursor = 0;
        while ($cursor < $offset) {
            $char = $document[$cursor];
            $cursor++;
            if ($char === "\r") {
                if ($cursor < $offset && $document[$cursor] === "\n") {
                    $cursor++;
                }
            } elseif ($char !== "\n") {
                continue;
            }
            $line++;
            $lineStart = $cursor;
        }

        $column = mb_strlen(substr($document, $lineStart, $offset - $lineStart), 'UTF-8') + 1;
        $lineLength = strcspn($document, "\r\n", $lineStart);
        $lineText = substr($document, $lineStart, $lineLength);

        return [$offset, $line, $column, $lineText];
    }

    /**
     * The offending line with a caret under the offending column, quoted the
     * way a compiler quotes source. A long line is shown as a window around
     * the caret, so the message stays the same size however long the line is.
     */
    private static function quote(int $number, string $lineText, int $column): string
    {
        [$quoted, $at] = self::windowAround($lineText, $column);
        $gutter = str_repeat(' ', strlen((string) $number));

        return "{$number} | {$quoted}\n{$gutter} | " . str_repeat(' ', $at - 1) . '^';
    }

    /**
     * Cut a line down to a window around the given column, and say which
     * column the offending character sits at in that window. Both columns
     * count from 1.
     *
     * @return array{0: string, 1: int}
     */
    private static function windowAround(string $lineText, int $column): array
    {
        $length = mb_strlen($lineText, 'UTF-8');
        if ($length <= self::QUOTED_LINE_WIDTH) {
            return [$lineText, $column];
        }

        $target = $column - 1;
        $lastStart = $length - self::QUOTED_LINE_WIDTH;
        $start = min(max($target - intdiv(self::QUOTED_LINE_WIDTH, 2), 0), $lastStart);
        $end = $start + self::QUOTED_LINE_WIDTH;

        $quoted = ($start > 0 ? self::ELLIPSIS : '')
            . mb_substr($lineText, $start, self::QUOTED_LINE_WIDTH, 'UTF-8')
            . ($end < $length ? self::ELLIPSIS : '');
        $shift = $start > 0 ? strlen(self::ELLIPSIS) : 0;

        return [$quoted, $target - $start + $shift + 1];
    }
}
